internal/provisioner/objects/rbac/rbac.go

```diff
@@ -53,10 +53,31 @@ func ensureContourRBAC(ctx context.Context, cli client.Client, contour *model.Co
 		return fmt.Errorf("failed to ensure service account %s/%s: %w", contour.Namespace, names.ServiceAccount, err)
 	}
 
-	// Ensure cluster role & binding.
-	if err := clusterrole.EnsureClusterRole(ctx, cli, names.ClusterRole, contour); err != nil {
-		return fmt.Errorf("failed to ensure cluster role %s: %w", names.ClusterRole, err)
+	// By default, Contour watches all namespaces, use default cluster role and rolebinding
+	clusterRoleForGatewayclassOnly := true
+	if contour.WatchAllNamespaces() {
+		// Ensure cluster role & binding.
+		if err := clusterrole.EnsureClusterRole(ctx, cli, names.ClusterRole, contour, !clusterRoleForGatewayclassOnly); err != nil {
+			return fmt.Errorf("failed to ensure cluster role %s: %w", names.ClusterRole, err)
+		}
+	} else {
+		// Ensure cluster role & cluster binding for gatewayclass first since it's cluster scope variables
+		if err := clusterrole.EnsureClusterRole(ctx, cli, names.ClusterRole, contour, clusterRoleForGatewayclassOnly); err != nil {
+			return fmt.Errorf("failed to ensure cluster role %s: %w", names.ClusterRole, err)
+		}
+
+		// Ensures role and rolebinding for set of namespaces in contour.spec.watchNamespaces variable
+		// Ensure cluster role & cluster binding for gatewayclass first since it's cluster scope variables
+		if err := role.EnsureRolesInNamespaces(ctx, cli, names.Role, contour, contour.Spec.WatchNamespaces); err != nil {
```

Comment:
Since we're reusing `names.Role` here (and `names.RoleBinding` for the role binding), there could be a conflict between this role/rolebinding and the "controller" role/rolebinding if `WatchNamespaces` contains the namespace where the Contour instance is deployed. I'm thinking it might be easiest to define new names to use for the per-namespace role/rolebinding (maybe `contour-resources-<instance name>`), and leave the "controller" role/rolebinding as-is, rather than trying to merge the roles together for the namespace where the Contour instance is deployed.